src/solver/cartesianprod/backward_cartesianprod.cpp

```diff
@@ -0,0 +1,167 @@
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include <miopen/conv_solution.hpp>
+#include <miopen/execution_context.hpp>
+#include <miopen/invoke_params.hpp>
+#include <miopen/tensor_view_utils.hpp>
+#include <miopen/cartesianprod/solvers.hpp>
+
+#include <miopen/cartesianprod/invoke_params.hpp>
+#include <miopen/datatype.hpp>
+#include <miopen/cartesianprod.hpp>
+#include <miopen/target_properties.hpp>
+#include <miopen/par_for.hpp>
+
+#define LOCAL_SIZE_BWD 256
+#define TILE_SIZE 16
+
+namespace miopen {
+
+namespace solver {
+
+namespace cartesianprod {
+
+bool IsOverRocmBwd(const miopen::cartesianprod::BwdProblemDescription& problem)
+{
+    for(size_t i = 0; i < problem.GetInputCount(); i++)
+    {
+        if(problem.GetInputGradDesc(i).GetElementSize() > 10)
+        {
+            return false;
+        }
+    }
+    return true;
+}
+
+bool CartesianProdBackward::IsApplicable(
+    const ExecutionContext&, const miopen::cartesianprod::BwdProblemDescription& problem) const
+{
+    if(!(problem.GetOutputGradDesc().GetType() == miopenHalf ||
+         problem.GetOutputGradDesc().GetType() == miopenFloat ||
+         problem.GetOutputGradDesc().GetType() == miopenBFloat16))
+    {
+        return false;
+    }
+    if(!problem.IsAllPacked())
+    {
+        return false;
+    }
+    if(!IsOverRocmBwd(problem))
+    {
+        return false;
+    }
+    return true;
+}
+
+ConvSolution CartesianProdBackward::GetSolution(
+    const ExecutionContext& context,
+    const miopen::cartesianprod::BwdProblemDescription& problem) const
+{
+    std::ignore = context;
+
+    auto dtype        = problem.GetOutputGradDesc().GetType();
+    auto output_dtype = miopen::GetDataType(dtype);
+    auto inputCount   = problem.GetInputCount();
+
+    auto build_params =
+        KernelBuildParameters{{"MIOPEN_USE_FP16", static_cast<int>(dtype == miopenHalf)},
+                              {"MIOPEN_USE_FP32", static_cast<int>(dtype == miopenFloat)},
+                              {"MIOPEN_USE_BFP16", static_cast<int>(dtype == miopenBFloat16)},
+                              {"TILE_SIZE", TILE_SIZE},
+                              {"D_TYPE", output_dtype == "bfloat16" ? "ushort" : output_dtype}};
+
+    if(inputCount == 1)
+    {
+        return ConvSolution{miopenStatusNotImplemented};
```

Comment:
We just have to copy ouput grad to input grad, right?

Reply:
added size 1 case